statsmodels/tsa/holtwinters.py

```diff
@@ -0,0 +1,726 @@
+"""
+Created on Wed Jul 12 09:35:35 2017
+
+Notes
+-----
+Code written using below textbook as a reference.
+Results are checked against the expected outcomes in the text book.
+
+Properties:
+Hyndman, Rob J., and George Athanasopoulos. Forecasting: principles and practice. OTexts, 2014.
+
+Author: Terence L van Zyl
+
+"""
+import numpy as np
+
+from statsmodels.base.model import Results
+from statsmodels.base.wrapper import populate_wrapper, union_dicts, ResultsWrapper
+from statsmodels.tsa.base.tsa_model import TimeSeriesModel
+
+from scipy.optimize import basinhopping, brute, minimize
+from scipy.spatial.distance import sqeuclidean
+try:
+    from scipy.special import inv_boxcox
+except ImportError:
+    def inv_boxcox(x, lmbda):
+        return np.exp(np.log1p(lmbda * x) / lmbda) if lmbda != 0 else np.exp(x)
+from scipy.stats import boxcox
+
+
+def _holt_init(x, xi, p, y, l, b):
+    p[xi] = x
+    alpha, beta, _, l0, b0, phi = p[:6];
+    alphac = 1 - alpha
+    betac = 1 - beta
+    y_alpha = alpha * y
+    l[:] = 0; b[:] = 0;
+    l[0] = l0
+    b[0] = b0
+    return alpha, beta, phi, alphac, betac, y_alpha
+
+
+def _holt__(x, xi, p, y, l, b, s, m, n, max_seen):
+    alpha, beta, phi, alphac, betac, y_alpha = _holt_init(x, xi, p, y, l, b)
+    for i in range(1, n):
+        l[i] = (y_alpha[i - 1]) + (alphac * (l[i - 1]))
+    return sqeuclidean(l, y)
+
+#(M,) (Md,)
+
+
+def _holt_mul_dam(x, xi, p, y, l, b, s, m, n, max_seen):
+    alpha, beta, phi, alphac, betac, y_alpha = _holt_init(x, xi, p, y, l, b)
+    if alpha == 0.0:
+        return max_seen
+    if beta > alpha:
+        return max_seen
+    for i in range(1, n):
+        l[i] = (y_alpha[i - 1]) + (alphac * (l[i - 1] * b[i - 1]**phi))
+        b[i] = (beta * (l[i] / l[i - 1])) + (betac * b[i - 1]**phi)
+    return sqeuclidean(l * b**phi, y)
+
+#(A,) (Ad,)
+
+
+def _holt_add_dam(x, xi, p, y, l, b, s, m, n, max_seen):
+    alpha, beta, phi, alphac, betac, y_alpha = _holt_init(x, xi, p, y, l, b)
+    if alpha == 0.0:
+        return max_seen
+    if beta > alpha:
+        return max_seen
+    for i in range(1, n):
+        l[i] = (y_alpha[i - 1]) + (alphac * (l[i - 1] + phi * b[i - 1]))
+        b[i] = (beta * (l[i] - l[i - 1])) + (betac * phi * b[i - 1])
+    return sqeuclidean(l + phi * b, y)
+
+
+def _holt_win_init(x, xi, p, y, l, b, s, m):
+    p[xi] = x
+    alpha, beta, gamma, l0, b0, phi = p[:6]; s0 = p[6:]
+    alphac = 1 - alpha
+    betac = 1 - beta
+    gammac = 1 - gamma
+    y_alpha = alpha * y
+    y_gamma = gamma * y
+    l[:] = 0; b[:] = 0; s[:] = 0
+    l[0] = l0; b[0] = b0; s[:m] = s0
+    return alpha, beta, gamma, phi, alphac, betac, gammac, y_alpha, y_gamma
+
+#(,M)
+
+
+def _holt_win__mul(x, xi, p, y, l, b, s, m, n, max_seen):
```

Comment:
add at least minimal docstrings also to private function
eg. """Recursion for multiplicative Holt-Winters"""

Reply:
done